rmw_fastrtps_shared_cpp/src/rmw_service.cpp

```diff
@@ -53,44 +50,103 @@ __rmw_destroy_service(
+  } else if (nullptr == participant_info || nullptr == common_context) {
+    RMW_SET_ERROR_MSG("destroy_service() called on invalid context");
+    final_ret = RMW_RET_INVALID_ARGUMENT;
+    info_is_valid = false;
+  }
+
+  if (info_is_valid) {
     // Update graph
     std::lock_guard<std::mutex> guard(common_context->node_update_mutex);
     rmw_gid_t gid = rmw_fastrtps_shared_cpp::create_rmw_gid(
-      identifier, info->request_subscriber_->getGuid());
+      identifier, info->request_reader_->guid());
     common_context->graph_cache.dissociate_reader(
       gid,
       common_context->gid,
       node->name,
       node->namespace_);
     gid = rmw_fastrtps_shared_cpp::create_rmw_gid(
-      identifier, info->response_publisher_->getGuid());
+      identifier, info->response_writer_->guid());
     rmw_dds_common::msg::ParticipantEntitiesInfo msg =
       common_context->graph_cache.dissociate_writer(
       gid, common_context->gid, node->name, node->namespace_);
-    ret = rmw_fastrtps_shared_cpp::__rmw_publish(
+    final_ret = rmw_fastrtps_shared_cpp::__rmw_publish(
       identifier,
       common_context->pub,
       static_cast<void *>(&msg),
       nullptr);
   }
 
-  Domain::removeSubscriber(info->request_subscriber_);
-  Domain::removePublisher(info->response_publisher_);
-  delete info->pub_listener_;
-  delete info->listener_;
-  _unregister_type(info->participant_, info->request_type_support_);
-  _unregister_type(info->participant_, info->response_type_support_);
-  delete info;
+  auto show_previous_error =
+    [&final_ret]() {
+      if (RMW_RET_OK != final_ret) {
+        RMW_SAFE_FWRITE_TO_STDERR(rmw_get_error_string().str);
+        RMW_SAFE_FWRITE_TO_STDERR(" during '" RCUTILS_STRINGIFY(__function__) "'\n");
+        rmw_reset_error();
+      }
+    };
+
+  /////
+  // Delete DataWriter and DataReader
+  if (info_is_valid) {
+    std::lock_guard<std::mutex> lck(participant_info->entity_creation_mutex_);
+
+    // Keep pointers to topics, so we can remove them later
+    auto response_topic = info->response_writer_->get_topic();
+    auto request_topic = info->request_reader_->get_topicdescription();
+
+    // Delete DataReader
+    ReturnCode_t ret = participant_info->subscriber_->delete_datareader(info->request_reader_);
+    if (ret != ReturnCode_t::RETCODE_OK) {
+      show_previous_error();
+      RMW_SET_ERROR_MSG("Fail in delete datareader");
+      final_ret = rmw_fastrtps_shared_cpp::cast_error_dds_to_rmw(ret);
+      info->request_reader_->set_listener(nullptr);
+    }
+
+    // Delete DataReader listener
+    if (nullptr != info->listener_) {
+      delete info->listener_;
+    }
+
+    // Delete DataWriter
+    ret = participant_info->publisher_->delete_datawriter(info->response_writer_);
+    if (ret != ReturnCode_t::RETCODE_OK) {
+      show_previous_error();
+      RMW_SET_ERROR_MSG("Fail in delete datawriter");
+      final_ret = rmw_fastrtps_shared_cpp::cast_error_dds_to_rmw(ret);
```

Comment:
@MiguelCompany it's important that return codes are restricted to those defined by the `rmw` API contract.